### ISSUE TITLE
Adding a module system similar to Java 9's JPMS 

### ISSUE THREAD

**Issue by CosminSontu** [dotnet/runtime#91670]
### Background and motivation

Background:
Usually when I develop modular monoliths, my modules are composed of several assemblies:
SomeModule.Ports.In.dll
SomeModule.Adaptors.In.dll
SomeModule.Core.dll
SomeModule.Storage.Out.MsSql.dll
SomeModule.Messaging.Out.VerneMq.dll

This is because there is no grouping concept between namespace and assembly, something like a Module that would be a physical artefact. Several modules could make up a dll.
I know .NET supports multi-file and sattelite assemblies but it's not quite the same. Tooling support (ex. Visual Studio) is also important for this feature.

Motivation:

Watching a recent ASP.NET Community Standup on AOT, I understood some .NET libs are refactored to be more modular in order to help trimming and that this new pattern will be applied for more parts of the framework.
I'm wondering if such feature would help making that pattern implementation seamless.

### API Proposal

//Java JPMS API can be a starting point

### API Usage

//JPMS API usage can be a starting point

**Comment by jkotas** [comment]
> I understood some .NET libs are refactored to be more modular in order to help trimming

Refactoring to help trimming have not involved splitting assemblies into multiple modules so far.

**Comment by jkotas** [comment]
> Java 9's JPMS

Existing .NET concepts (Nuget packages, reference assemblies, implementation assemblies, etc.) provide similar capabilities as JPMS.

Could you please explain the problem that you are trying to solve by this proposal in more details?

**Comment by CosminSontu** [comment]
Thanks for your answer.
Basically whenever I am defining a "module" in the context of a modular monolith, this "module" is composed of a group of assemblies as listed above.

I can see how Reference Assemblies could make some of those assemblies (the ones containing contracts), dissapear. 
My module will be composed of a cluster of assemblies in the end:

SomeModule.Ports.In.dll
SomeModule.Core.dll
SomeModule.Storage.Out.MsSql.dll
SomeModule.Messaging.Out.VerneMq.dll

What I'm looking for, is to have the ability that my "module" is a single physical assembly:
SomeModule.dll
And I would liketo have another physical level below the assembly (I will confusingly use the term "netmodule") since namespaces are a weak boundary from my point of view:
SomeModule.dll  
     |
    + - SomeModule.Ports.In.netmodule
    + - SomeModule.Core.netmodule
    + - SomeModule.Storage.Out.MsSql.netmodule
    + - SomeModule.Messaging.Out.VerneMq.netmodule

Ideally there would be project files specific to netmodules since they can have their own dependencies.

The project file corresponding to SomeModule.dll  (our "module") would summarize the dependencies of included netmodules (if one excludes  a netmodule, the summary would be udpated accordingly)
This project file would also control visibility of contained netmodules contracts.
At compile time, SomeModule.dll will be built by **linking** the included netmodules.

Basically, I am missing some kind physical partitioning of assemblies and the only way I can do this today in .NET is using a folder structure and a set of related projects / assemblies.

**Comment by jkotas** [comment]
> the only way I can do this today in .NET is using a folder structure and a set of related projects / assemblies.

Yes, it is fairly common to organize source code into a folder structure for large assemblies. We do that a lot in this repo. Why is that not good enough?

**Comment by jkotas** [comment]
You can setup the enforcement of internal architecture by building the same sources twice (using two sets of project files). One set of projects builds large assemblies that you ship in nuget packages. Second set of projects builds fine grained assemblies to enforce your internal architecture boundaries. (BTW: WPF was built like that in the past to enforce internal architecture layering.)

**Comment by CosminSontu** [comment]
I'll give it a try. Thanks!
